File: Python/Sorted_Squared_Array.py
# Time: O(n log(n))
# Space: O(n)
def sortedSquaredArray(array):
    ans = []

    for i in array:
        x = i * i
        ans.append(x)

    ans.sort()
    return ans


# Time: O(n)
# Space: O(n)
def sortedSquaredArray2(array):
    sortedSquares = [0 for _ in array]
    smallerValueIdx = 0
    largerValueIdx = len(array) - 1

    for idx in reversed(range(len(array))):
        smallerValue = array[smallerValueIdx]
        largerValue = array[largerValueIdx]

        if abs(smallerValue) > abs(largerValue):
            sortedSquares[idx] = smallerValue * smallerValue
            smallerValueIdx += 1
        else:
            sortedSquares[idx] = largerValue * largerValue
            largerValueIdx -= 1

    return sortedSquares

File: Java/Sorted_Squared_Array.java
package Java;

import java.util.*;

// Time: O(n log(n))
// Space: O(n)
class Program {

    public int[] sortedSquaredArray(int[] array) {
        int[] list = new int[array.length];

        for (int i = 0; i < array.length; i++) {
            int x = array[i] * array[i];
            list[i] = x;
        }

        Arrays.sort(list);
        return list;
    }
}

class Program2 {
    public int[] sortedSquaredArray(int[] array) {
        int[] sortedArray = new int[array.length];
        int smallerValueIdx = 0;
        int largerValueIdx = array.length - 1;

        for (int i = array.length - 1; i >= 0; i--) {
            int smallerValue = array[smallerValueIdx];
            int largerValue = array[largerValueIdx];

            if (Math.abs(smallerValue) > Math.abs(largerValue)) {
                sortedArray[i] = smallerValue * smallerValue;
                smallerValueIdx++;
            } else {
                sortedArray[i] = largerValue * largerValue;
                largerValueIdx--;
            }
        }
        return sortedArray;
    }
}